### PR TITLE
Avoid duplicate conditions when using #or

### PR DESCRIPTION
**Pull request by MaxLap**
Right now, when using `#or`, nothing is done to try to avoid duplicated conditions. As a result, query length increases at an exponential rate. This makes it really hard to parse for humans, making debugging / understanding the query log really hard.

This PR does basic work to extract the predicates and binds that are common to both associations beings `#or`-ed and put them first, before the `#or`. 

To show the scale of the problem:
```
class Project
    scope :big, -> { where(big1: true).or(where(big2: true)) }
    scope :important, -> { where(important1: true).or(where(important2: true)) }
    scope :ongoing, -> { where(ongoing1: true).or(where(ongoing2: true)) }
end

Project.big # 2 conditions
#=> SELECT...WHERE ("projects"."big1" = ? OR "projects"."big2" = ?) 
Project.big.important # 6 conditions
#=> SELECT...WHERE (("projects"."big1" = ? OR "projects"."big2" = ?) AND "projects"."important1" = ? OR ("projects"."big1" = ? OR "projects"."big2" = ?) AND "projects"."important2" = ?)
Project.big.important.ongoing # 14 conditions
#=> To long to show
# Then 30 conditions
# Then 62
# Then 126 conditions, for using #or 6 times. This should be 12!

# After this fix, it is correctly 2, 4, 6, 8, 10, ... conditions.
```

To make further development in WhereClause easier, the basic behavior of iterating over the predicate and binds has been generalized.

In the current code, there are 2 questions i ask as comments:
* I wonder if `#+` should be changed to do the same behavior as `#|`. The only reason not to, in my mind, is if someone considers this to be a performance issue.
* `#partition_common_left_right` is currently less readable than it could be if i just used the new operators I added to WhereClause. However, those operators might be a little slower since it means more iterations over the lists in the WhereClauses.

Because the the logic for finding the binds that go with a predicate has been extracted and fixed, this also fixes the #29780. However, the test case in #29780 is not included in this PR, and I assume #29780 will be backported, unlike this PR.

### CODE REVIEW

**Comment by MaxLap** [comment]
The failed check seems to betravis-ci getting an IOError for one of its builds.

**Comment by sgrif** [comment]
This was originally part of the implementation of `#or`. It was removed as it was deemed to add too little value for the complexity it adds to the code base.

**Comment by MaxLap** [comment]
@sgrif: The only reason there is any complexity in this commit is because of how barebone the current implementation of WhereClause is. This implementation forces any alteration to a WhereClause to need to iterate the whole ast and keep track of the binds, whileno helper is provided to help with this, leading to duplication of this logic.

This PR adds tools so that this iterating behavior can be reused, instead of being copy pasted in every PR that has to do a modification and uses said tool on the existing code that needs this iteration.

If you look only at #or, this is you will see what is the total amount of added complexity. 

Exponential conditions makes the generated SQL absolutely un readable and undebugable. You can't even understand what the code is trying to do when you see SQL that has 100+ clause where is should have maybe 12 to 15. It forces people to use workaround, such as using `rel.merge(rel.where(foo: 1).or(rel.where(bar:1))` instead of just using `rel.where(foo: 1).or(rel.where(bar: 1))`

Here is the difference from calling 3 scopes in a row that each do #or, online formatter used because without newlines, that's just way too long. (`Project.big.software.important.to_sql`)

```
SELECT "projects".*
FROM   "projects"
WHERE  ( ( ( "projects"."big1" = 't'
              OR "projects"."big2" = 't' )
           AND "projects"."software1" = 't'
            OR ( "projects"."big1" = 't'
                  OR "projects"."big2" = 't' )
               AND "projects"."software2" = 't' )
         AND "projects"."important1" = 't'
          OR ( ( "projects"."big1" = 't'
                  OR "projects"."big2" = 't' )
               AND "projects"."software1" = 't'
                OR ( "projects"."big1" = 't'
                      OR "projects"."big2" = 't' )
                   AND "projects"."software2" = 't' )
             AND "projects"."important2" = 't' )  

VS

SELECT "projects".*
FROM   "projects"
WHERE  ( "projects"."big1" = 't'
          OR "projects"."big2" = 't' )
       AND ( "projects"."software1" = 't'
              OR "projects"."software2" = 't' )
       AND ( "projects"."important1" = 't'
              OR "projects"."important2" = 't' )  
```
I don't consider this to "add too little", it removes a lot of noice. If I added one more such scope, the first query would grow by double and add another layer of nesting, while the second would grow by 33% only without any nesting.

**Comment by MaxLap** [comment]
@sgrif, Can I have some feedback or answer to my comment? Has there been any discussion about this issue somewhere?

I've split the commit in two and reworked it a little to make it less complex and show the first step of extracting the logic and the second one of adding the fix (in my mind, this is a bug).

I beleive those changes don't show in this PR because it was closed. Here is a link to the commits: https://github.com/MaxLap/rails/commits/avoid_or_clause_duplicates

Thank you

**Comment by MaxLap** [comment]
Hey @sgrif, following your rework of WhereClause, the code needed for this feature became really simple.

I made 2 versions of the PR, you can find the commits here:

* v1: https://github.com/MaxLap/rails/commit/d30d9fe3222443e3b95ad49bf9cebca9c0ffd15d
* v2: https://github.com/MaxLap/rails/commit/82bada0fc7d1a1d53cc6b5397dcad292c02e8758

Is any of those acceptable?

**Comment by sgrif** [comment]
First one seems fine. The test is a little gnarly. If you can make the test more clear what's going on I think it's fine. If you want to open a new PR we can discuss there.
